### PR TITLE
chore: send device diagnostic metrics

### LINKED ISSUE #459: Gather device diagnostic metrics once a month

### PR DESCRIPTION
**Pull request by EvanHahn**
We've changed our plan for metrics.

This implements one part of the new plan: **device diagnostic metrics**. This includes device manufacturer, device type, total RAM, and other data about the device itself.

These metrics should be sent once a month, as they are unlikely to change.

_Notes for reviewers:_

- I don't think it's really worth looking at the deleted files. They represent our old plan.
- I would start in `useSendMetrics.ts`, then go to `useSendDeviceDiagnosticMetrics.ts`, and follow the code from there.

Closes #459.

### CODE REVIEW

File: src/frontend/__mocks__/expo-crypto.ts
Comment (EvanHahn) [inline src/frontend/__mocks__/expo-crypto.ts:1]: Jest can't import `expo-crypto`. This mocks out the parts we use so tests pass.

File: src/frontend/hooks/persistedState/usePersistedMetricsDeviceId.ts
Comment (EvanHahn) [inline src/frontend/hooks/persistedState/usePersistedMetricsDeviceId.ts:1]: This is a persisted "metrics device ID", stored for the lifetime of the app. I couldn't figure out a better way to do this, but maybe there's some stable device ID we can grab from the OS somehow.

(I found something for Android but not for iOS.)

File: src/frontend/metrics/getMonthlyHash.ts
Comment (EvanHahn) [inline src/frontend/metrics/getMonthlyHash.ts:1]: This is currently just used for device diagnostic metrics, but we also plan to use it in a few other places.

File: src/frontend/metrics/getMonthlyHash.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  writeStringToBytes(result, 3 + salt.length, metricsDeviceId);
+  return result;
+}
+
+export async function getMonthlyHash(
+  salt: string,
+  metricsDeviceId: string,
+  now: Date,
+): Promise<string> {
+  const hashedBuffer = await digest(
```
Comment (gmaclennan) [inline src/frontend/metrics/getMonthlyHash.ts:39]: What benefit does this have over:

```ts
return await digestStringAsync(
  CryptoDigestAlgorithm.SHA512,
  `${now.getUTCFullYear()}-${now.getUTCMonth()}-${metricsDeviceId}`
)
```
?

File: src/frontend/metrics/deviceDiagnosticMetrics.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import pProps from 'p-props';
+import {Dimensions, PixelRatio, Platform} from 'react-native';
+import DeviceInfo from 'react-native-device-info';
```
Comment (gmaclennan) [inline src/frontend/metrics/deviceDiagnosticMetrics.ts:3]: Just highlighting that [expo-device](https://docs.expo.dev/versions/latest/sdk/device/) does, I think, more normalization of this data between platforms, and also provides more data as constants rather than async. We are however already using react-native-device-info in a few places, so fine to use here for now. As we add the daily app data, it may be more convenient to switch to expo-device.

File: src/frontend/metrics/getMonthlyHash.ts (hunk 1 above)
Comment (EvanHahn) [reply src/frontend/metrics/getMonthlyHash.ts]: Good point. Simplified in d08e69d8727f4b7ff6114317688ec42fdc5f8ff2.

File: src/frontend/metrics/deviceDiagnosticMetrics.ts (hunk 2 above)
Comment (EvanHahn) [reply src/frontend/metrics/deviceDiagnosticMetrics.ts]: The fact that it returns data synchronously makes things much easier. Switched in a5b55f246c9a7f5eecd2ace1fdeb2ceb25d986e8.

**Comment by EvanHahn** [comment]
Rewrote this to use a singleton.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      return 'TV';
+    case null:
+    case Device.DeviceType.UNKNOWN:
+    default:
+      return 'UNKNOWN';
+  }
+}
+
+function getMetricsDeviceId(): string {
+  const result = storage.getString('MetricsDeviceId');
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:58]: I think this should be read/set at initialization, and kept in memory. Storage is fast, but it still has a cost, and there is no need to re-read this every time.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  if (isEmulator) result.isEmulator = true;
+
+  return result;
+}
+
+const isAppActive = (appState: AppStateStatus): boolean =>
+  appState === 'active';
+
+const hasEnoughTimeElapsed = (): boolean => {
+  const lastSentAtMs = storage.getNumber(STORAGE_KEY) ?? -Infinity;
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:113]: As above, read at initialization, and keep the result in memory.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    this.#update();
+  }
+
+  async #update(): Promise<void> {
+    this.#updateSubscriptions();
+
+    const shouldSendMetrics =
+      this.#isEnabled &&
+      this.#isOnline &&
+      this.#isAppActive &&
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:145]: Read this from AppState.currentState rather than storing it as state - there is no additional cost to reading this, and keeping the two states in sync just makes things difficult for ourselves (e.g. currently there are edge-cases where this could not be the same as `AppState.currentState` if setEnabled() is called at certain times in the app lifecycle, or the class is initialized at certain times).

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+  #subscriptionCleanupFns: null | (() => void)[] = null;
+
+  setEnabled(isEnabled: boolean): void {
+    this.#isEnabled = isEnabled;
+    this.#update();
+  }
+
+  async #update(): Promise<void> {
+    this.#updateSubscriptions();
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:140]: Personal opinion: Updating these as a side-effect of setting enabled (vs. just subscribing / unsubscribing within setEnabled) creates more scope for introducing bugs (e.g. circular code). I don't see the advantage of having these as a side-effect rather than within enable/disable action.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+      const appStateSubscription = AppState.addEventListener(
+        'change',
+        appState => {
+          this.#isAppActive = isAppActive(appState);
+          this.#update();
+        },
+      );
+      subscriptionCleanupFns.push(() => appStateSubscription.remove());
+
+      const nextMonthTimeout = setLongTimeout(() => {
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:187]: I can't find any info online about leaving long timeouts when the app is in the background, however I feel like we should avoid doing that, in case it results in either the timeout being cleared (which will lead to unexpected behaviour), or the app being closed by the OS. I think it's safest to just run this while the app is in the foreground (active state).

Also, setLongTimeout is very clever, however if we are only handling the use-case of the app being in the foreground, the main edge-case we are trying to catch with the timeout is the user having the app open and connected over the month-end. It may be worth just removing this, and acknowledging that if a user keeps the app open for 25 days without turning off their phone or switching to another app, then metrics will fail to send, and that avoids adding additional code to maintain.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  );
+  setIfNotNull(result, 'totalMemory', Device.totalMemory);
+
+  const isEmulator = !Device.isDevice;
+  if (isEmulator) result.isEmulator = true;
+
+  return result;
+}
+
+const isAppActive = (appState: AppStateStatus): boolean =>
```
Comment (gmaclennan) [inline src/frontend/metrics/DeviceDiagnosticMetrics.ts:109]: Complete personal preference: just write `appState === 'active'` in the code. Adding this helper function means that a reviewer / maintainer needs to look up what this helper does, and it doesn't really save many characters.

File: src/frontend/hooks/useSendMetrics.ts
Comment (gmaclennan) [inline src/frontend/hooks/useSendMetrics.ts:1]: Is this used anywhere any more?

File: src/frontend/hooks/persistedState/createPersistedState.ts
Comment (gmaclennan) [inline src/frontend/hooks/persistedState/createPersistedState.ts:1]: Are these changes still necessary?

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 7 above)
Comment (EvanHahn) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts]: Removed in 7deb5d846d7365b24a46cac61005bed10c2174e8.

File: src/frontend/hooks/useSendMetrics.ts
Comment (EvanHahn) [reply src/frontend/hooks/useSendMetrics.ts:1]: Nope, thanks for catching. Removed in 94c925e93ff5e9ab56fd5fddda5eb74fabcee009.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 5 above)
Comment (EvanHahn) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts:135]: Changed in f94bdc53aca2fee09b3144baffd35805233f051b.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 8 above)
Comment (EvanHahn) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts]: Removed in f94bdc53aca2fee09b3144baffd35805233f051b.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 3 above)
Comment (EvanHahn) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts:53]: I'd rather read it every time, especially given that this is called infrequently. Keeping the two states in sync is error-prone; for example, if someone sets the key outside this class, the in-memory value will be wrong.

I don't feel strongly, but I'm going to leave things as is. Let me know if you want to profile this or change it.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 6 above)
Comment (EvanHahn) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts]: Moved in d8830d5cfedede2a8116ccd35b1619355fa6b2aa.

File: src/frontend/hooks/persistedState/createPersistedState.ts
Comment (EvanHahn) [reply src/frontend/hooks/persistedState/createPersistedState.ts:1]: No. Reverted in 8db0c255cb326a6f8252c8621b9be422724cc33c.

File: src/frontend/metrics/DeviceDiagnosticMetrics.ts (hunk 4 above)
Comment (gmaclennan) [reply src/frontend/metrics/DeviceDiagnosticMetrics.ts:99]: agreed about the out-of-sync risk, and for deviceId I guess it's not called often, however one would be called every time the user gets a connection, and every time the app opens, and modifying this value outside this class would break things anyway, however the perf impact is relatively small, just want to flag this to keep an eye on it if we do start seeing perf issues - I think mmkv is fairly fast, but I've seen issues in the past with storage causing slowdowns in react native apps on older / lower cost devices.

**Review by gmaclennan — APPROVED** [review]
